jgrapht-core/src/main/java/org/jgrapht/alg/CliqueMinimalSeparatorDecomposition.java

```diff
@@ -0,0 +1,452 @@
+ *                   Tomas Hruz
+ *                   Philipp Hoppen
+ *
+ * $Id$
+ *
+ * Changes
+ * -------
+ * 06-Feb-2015 : Initial revision (FB);
+ *
+ */
+package org.jgrapht.alg;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+
+import org.jgrapht.Graph;
+import org.jgrapht.Graphs;
+import org.jgrapht.graph.SimpleGraph;
+import org.jgrapht.graph.Subgraph;
+
+/**
+ * Clique Minimal Separator Decomposition using MCS-M+ and Atoms algorithm as
+ * described in Berry et al. An Introduction to Clique Minimal Separator
+ * Decomposition (2010), DOI:10.3390/a3020197, 
+ * <a href="http://www.mdpi.com/1999-4893/3/2/197">
+ * http://www.mdpi.com/1999-4893/3/2/197</a>
+ * 
+ * <p>
+ * The Clique Minimal Separator (CMS) Decomposition is a procedure that splits a
+ * graph into a set of subgraphs separated by minimal clique separators, adding
+ * the separating clique to each component produced by the separation. At the
+ * end we have a set of atoms. The CMS decomposition is unique and yields the
+ * set of the atoms independent of the order of the decomposition.
+ * 
+ * @author Florian Buenzli <fbuenzli@student.ethz.ch>
+ * @author Thomas Tschager <thomas.tschager@inf.ethz.ch>
+ * @author Tomas Hruz <tomas.hruz@inf.ethz.ch>
+ * @author Philipp Hoppen
+ */
+public class CliqueMinimalSeparatorDecomposition<V, E>
+{
+
+    /** Source graph to operate on */
+    private SimpleGraph<V, E> graph;
+
+    /** Minimal triangulation of graph */
+    private SimpleGraph<V, E> chordalGraph;
+
+    /** Fill edges */
+    private Set<E> fillEdges;
+
+    /** Minimal elimination ordering on the vertices of graph */
+    private LinkedList<V> meo;
+
+    /**
+     * List of all vertices that generate a minimal separator of
+     * <code>chordGraph</code>
+     */
+    private List<V> generators;
+
+    /** Set of clique minimal separators */
+    private Set<Set<V>> separators;
+
+    /** The atoms generated by the decomposition */
+    private Set<Set<V>> atoms;
+
+    /** Map for each separator how many components it produces. */
+    private Map<Set<V>, Integer> fullComponentCount = 
+        new HashMap<Set<V>, Integer>();
+
+    /**
+     * Setup a clique minimal separator decomposition on graph <code>g</code>.
+     * 
+     * @param g The graph to decompose.
+     */
+    public CliqueMinimalSeparatorDecomposition(SimpleGraph<V, E> g)
+    {
+        this.graph = g;
+        this.fillEdges = new HashSet<E>();
+    }
+
+    /**
+     * Compute the minimal triangulation of the graph. Implementation of
+     * Algorithm MCS-M+ as described in Berry et al. (2010), DOI:10.3390/a3020197
+     * <a href="http://www.mdpi.com/1999-4893/3/2/197">
+     * http://www.mdpi.com/1999-4893/3/2/197</a>
+     */
+    public void computeMinimalTriangulation()

```

Comment:
This should probably be private since the public methods which need it call it lazily.